kafka.go

```diff
@@ -1,161 +1,213 @@
-		configSync = sarama.NewConfig()
-	default:
-		configAsync = sarama.NewConfig()
-		configSync = sarama.NewConfig()
-	}
-
-	asyncProducer, err := sarama.NewAsyncProducer(brokerList, configAsync)
-	if err != nil {
-		return nil, err
+func setConfig(writerConfig *kafka.WriterConfig, readerConfig *kafka.ReaderConfig, config *KafkaConfig) {
+
+	if config.ReadTimeout != 0 {
+		writerConfig.ReadTimeout = config.WriteTimeout
+	}
+	if config.WriteTimeout != 0 {
+		writerConfig.WriteTimeout = config.WriteTimeout
 	}
+	if config.DialTimeout != 0 {
+		dialer := &kafka.Dialer{
+			Timeout: config.DialTimeout,
+		}
+		writerConfig.Dialer = dialer
+		readerConfig.Dialer = dialer
+	}
+}
 
-	configSync.Producer.Return.Successes = true
-	syncProducer, err := sarama.NewSyncProducer(brokerList, configSync)
-	if err != nil {
-		return nil, err
+// NewKafkaClient create a new instance of KafkaClient
+func NewKafkaClient(brokers []string, realm string, config ...*KafkaConfig) (*KafkaClient, error) {
+
+	writerConfig := &kafka.WriterConfig{
+		Brokers:  brokers,
+		Balancer: &kafka.LeastBytes{},
 	}
 
-	go listenToError(asyncProducer)
+	readerConfig := &kafka.ReaderConfig{
+		Brokers:  brokers,
+		MaxBytes: 10e6, // 10MB
+	}
+
+	// set client configuration
+	if len(config) > 0 {
+		setConfig(writerConfig, readerConfig, config[0])
+	}
 
 	return &KafkaClient{
-		realm:         realm,
-		asyncProducer: asyncProducer,
-		syncProducer:  syncProducer,
+		realm:           realm,
+		publishConfig:   *writerConfig,
+		subscribeConfig: *readerConfig,
+		subscribeMap:    make(map[string]map[string]func(*Event, error)),
 	}, nil
 }
 
-func listenToError(producer sarama.AsyncProducer) {
-	for err := range producer.Errors() {
-		logrus.Error("unable to push event to Kafka stream: ", err)
+// Publish send event to message broker
+func (client *KafkaClient) Publish(publishBuilder *PublishBuilder) {
+	message, err := constructEvent(publishBuilder)
+	if err != nil {
+		logrus.Errorf("unable to construct event : %s , error : %v", publishBuilder.eventName, err)
+		return
 	}
-}
 
-type eventEntry struct {
-	Event
+	for _, topic := range publishBuilder.topic {
+		config := client.publishConfig
+		config.Topic = topic
+		writer := kafka.NewWriter(config)
 
-	encoded []byte
-	err     error
+		err = writer.WriteMessages(publishBuilder.ctx, message)
+		if err != nil {
+			logrus.Errorf("unable to publish event to kafka. topic : %s , error %v : ", topic, err)
+		}
+	}
 }
 
-func (evtEntry *eventEntry) ensureEncoded() {
-	if evtEntry.encoded == nil && evtEntry.err == nil {
-		evtEntry.encoded, evtEntry.err = json.Marshal(evtEntry)
+// constructEvent construct event message
+func constructEvent(publishBuilder *PublishBuilder) (kafka.Message, error) {
+	id := generateID()
+	event := &Event{
+		ID:        id,
+		EventName: publishBuilder.eventName,
+		Namespace: publishBuilder.namespace,
+		ClientID:  publishBuilder.clientID,
+		UserID:    publishBuilder.userID,
+		Timestamp: time.Now(),
```

Comment:
don't you want to use UTC time?

Reply:
you're right, it should be in UTC
will update it thanks